File: cpp/tree/1991.cpp
#include <iostream>
#include <vector>
#include <string>
#define MAX 26

using namespace std;

struct node{
    char left;
    char right;
};

void preOrder(char node);
void inOrder(char node);
void postOrder(char node);

vector<node> tree(MAX);

int main() {
    char root;
    char left;
    char right;
    int N;
    cin >> N;
    for (int i = 0; i < N; i++) {
        cin >> root >> left >> right;
        tree[root].left = left;
        tree[root].right = right;
    }
    preOrder('A');
    cout << endl;
    inOrder('A');
    cout << endl;
    postOrder('A');
}


void preOrder(char node){
    if (node == '.') return;
    cout << node;
    preOrder(tree[node].left);
    preOrder(tree[node].right);
}

void inOrder(char node){
    if (node == '.') return;
    inOrder(tree[node].left);
    cout << node;
    inOrder(tree[node].right);
}

void postOrder(char node){
    if (node == '.') return;
    postOrder(tree[node].left);
    postOrder(tree[node].right);
    cout << node;
}

File: py/tree/1991.py
# 백준 1991

N = int(input())


tree = [None] * (2**(N-1))

for _ in range(N):
    root, left, right = input().split()
    if root == 'A':
        tree[0] = root
    root_index = tree.index(root)
    if left != '.':
        tree[root_index * 2 + 1] = left
    if right != '.':
        tree[root_index * 2 + 2] = right

def preorder(tree, i=0):
    if i < len(tree):
        print(tree[i], end = "")
        left = 2 * i + 1
        right = left + 1
        if left < len(tree) and tree[left] is not None:
            preorder(tree, left)
        if right < len(tree) and tree[right] is not None:
            preorder(tree, right)

def inorder(tree, i=0):
    if i < len(tree) or tree[i] is not None:
        left = 2 * i + 1
        right = left + 1
        if left < len(tree) and tree[left] is not None:
            inorder(tree, left)
        print(tree[i], end="")
        if right < len(tree) and tree[right] is not None:
            inorder(tree, right)

def postorder(tree, i=0):
    if i < len(tree) or tree[i] is not None:
        left = 2 * i + 1
        right = left + 1
        if left < len(tree) and tree[left] is not None:
            postorder(tree, left)
        if right < len(tree) and tree[right] is not None:
            postorder(tree, right)
        print(tree[i], end="")


preorder(tree)
print("")
inorder(tree)
print("")
postorder(tree)